src/metabase/lib/util.cljc

```diff
@@ -0,0 +1,100 @@
+(ns metabase.lib.util
+  "TODO -- this should probably be rolled into [[metabase.lib.query]]."
+  (:require
+   [clojure.set :as set]
+   [metabase.lib.options :as lib.options]
+   [metabase.lib.schema :as lib.schema]
+   [metabase.lib.schema.common :as lib.schema.common]
+   [metabase.shared.util.i18n :as i18n]
+   [metabase.util.malli :as mu]))
+
+(defn- native-query->pipeline [query]
+  (merge {:lib/type :mbql/query
+          :type     :pipeline
+          ;; we're using `merge` here instead of threading stuff so the `:lib/` keys are the first part of the map for
+          ;; readability in the REPL.
+          :stages   [(merge (lib.options/ensure-uuid {:lib/type :mbql.stage/native})
+                            (set/rename-keys (:native query) {:query :native}))]}
+         (dissoc query :type :native)))
+
+(defn- inner-query->stages [{:keys [source-query], :as inner-query}]
+  (let [stages     (if source-query
+                     (inner-query->stages source-query)
+                     [])
+        stage-type (if (:native inner-query)
+                     :mbql.stage/native
+                     :mbql.stage/mbql)
+        ;; we're using `merge` here instead of threading stuff so the `:lib/` keys are the first part of the map for
+        ;; readability in the REPL.
+        this-stage (merge (lib.options/ensure-uuid
+                           {:lib/type stage-type})
+                          (dissoc inner-query :source-query))]
+    (conj stages this-stage)))
+
+(defn- mbql-query->pipeline [query]
+  (merge {:lib/type :mbql/query
+          :type     :pipeline
+          :stages   (inner-query->stages (:query query))}
+         (dissoc query :type :query)))
+
+(mu/defn pipeline :- ::lib.schema/query
+  "Take a 'traditional' MBQL query and convert it to a `:pipeline` query."
```

Comment:
Well, it can "convert" pipeline queries too...

Reply:
Fixed the docstring